File: message.py
import random

class Message():
    def __init__(self, name, pid):
        self.name = name
        self.pid = pid

    def __str__(self):
        return self.name

    def __eq__(self, other):
        if (self.name == other.name):
            return True
        return False

class DummyMessage(Message):
    def __init__(self):
        pid = random.getrandbits(32)
        Message.__init__(self, "R%04x" % pid, pid)

class RealMessage(Message):
    def __init__(self, nid):
        pid = random.getrandbits(32)
        Message.__init__(self, "M%d/%04x" % (nid, pid), pid)

class EvilMessage(Message):
    def __init__(self, nid):
        pid = random.getrandbits(32)
        Message.__init__(self, "E%d/%04x" % (nid, pid), pid)

class RLC(Message):
    def __init__(self, messages):
        coefs = [random.randint(0, 255) for p in messages]
        pid = random.getrandbits(32)
        name = "LC/(" + ",".join([p.name for p in messages]) + ")" + "/(" + \
                       ",".join(["%02x" % c for c in coefs]) + ")"
        Message.__init__(self, name, pid)
        self.messages = messages
        self.coefs = coefs

File: src-lua/message.lua
local message = {}

local Message = {}
Message.__index = Message
setmetatable(Message, {__call = function(self, ...) return self.new(...) end})

function Message.new(nid, mid, name)
    local self = setmetatable({nid = nid, mid = mid, name = name}, Message)
    return self
end

function Message:__eq(other)
    return self.mid == other.mid
end

function Message:__tostring()
    return self.name
end

local DummyMessage = setmetatable({}, {__call = function(self, ...) return self.new(...) end})

function DummyMessage.new(nid)
    local mid = math.random(0, 2^32-1)
    local name = string.format("D%d-%04x", nid, mid)
    return Message.new(nid, mid, name)
end

local RealMessage = setmetatable({}, {__call = function(self, ...) return self.new(...) end})

function RealMessage.new(nid)
    local mid = math.random(0, 2^32-1)
    local name = string.format("R%d-%04x", nid, mid)
    return Message.new(nid, mid, name)
end

local EvilMessage = setmetatable({}, {__call = function(self, ...) return self.new(...) end})

function EvilMessage.new(nid)
    local mid = math.random(0, 2^32-1)
    local name = string.format("E%d-%04x", nid, mid)
    return Message.new(nid, mid, name)
end

local RLCMessage = setmetatable({}, {__call = function(self, ...) return self.new(...) end})

function RLCMessage.new(nid, messages)
    local mid = math.random(0, 2^32-1)

    local coefs = {}
    for i=1,#messages do
        coefs[i] = math.random(0, 2^8-1)
    end

    local name = string.format("LC%d-%04x", nid, mid)

    local self = Message.new(nid, mid, name)
    self.coefficients = coefs
    self.messages = messages
    return self
end

message.Message = Message
message.DummyMessage = DummyMessage
message.RealMessage = RealMessage
message.EvilMessage = EvilMessage
message.RLCMessage = RLCMessage

return message
